fix wrong class name (#256)
apache/incubator-dubbo-website: docs/en-us/dev/SPI.md

## SPI Config

### Source:

Dubbo SPI is inherited from standard JDK SPI(Service Provider Interface) and makes it more powerful.

Dubbo fixed below issues of the standard JDK SPI:

* the standard JDK SPI will load and instantize all implementation at once. It will be a waste of resources if the implementation is timecosted, but never be used.
* We cann't accquire the SPI name, if loading the SPI implementation is failed.For example: standard JDK ScriptEngine, get script type by invoking method getName(). RubyScriptEngine class will load failed if the depenency jar jruby.jar is missing, and the real error info will be lost. When user executes ruby scripts, program throws exception, telling not support ruby, but it is not the real cause.
* Enhance the SPI functionality by supporting IoC and AOP, one SPI can be easily injected by another SPI simply using setter.

### Appointment:

In the jar file containing extension class [^1], places a config file `META-INF/dubbo/full interface name`, file content pattern: `SPI name=the fully qualified name for the extension class`, use new line seperator for multiple implementation.

### Example:

> To extend Dubbo Protocol, placee a text file in the extension jar file: `META-INF/dubbo/org.apache.dubbo.rpc.Protocol`, content:

Author: No need for `please`

> To extend Dubbo Protocol, placee a text file in the extension jar file: `META-INF/dubbo/org.apache.dubbo.rpc.Protocol`, content:

Reviewer: > No need for `please`

ok

}
```

when `ExtensionLoader` loads `CarMaker`'s implementation `RaceCarMaker`, the method `setWheelMaker` needs paramType `WheelMaker` which is also a SPI, It will be automatically loaded.

This brings a new question:How `ExtensionLoader` determines which implementation to use when load the injected SPI. As for this demo, when existing multi `WheelMaker` implementation, which one should the `ExtensionLoader` chooses.

Good question, we will explain it in the following chapter: SPI Auto Adaptive.

### SPI Auto Adaptive

> The depency SPI that `ExtensionLoader` injects is an instance of `Adaptive`, the real spi implementation is known until the adaptive instance is be executed.

Reviewer: ```suggestion
The depency SPI that `ExtensionLoader` injects is an instance of `Adaptive`, the real spi implementation is known until the adaptive instance is executed.
```

Dubbo use URL (containing Key-Value) to pass the configuration.

The SPI method invocation has the URL parameter（Or Entity that has URL attribute）

> In this way depended SPI can get configuration from URL, after config all SPI key needed, configuration information will be passed from outer by URL.URL act as a bus when pass the config information.

Reviewer: ```suggestion
In this way depended SPI can get configuration from URL, after config all SPI key needed, configuration information will be passed from outer by URL. URL acts as a bus when passing the config information.
```

}
```

> when`ExtensionLoader` loading `CarMaker`  implementation `RaceCarMaker` ，`setWheelMaker`  needs paramType `WheelMaker`  which is also an SPI, It will be automatically loaded .

Reviewer: ```suggestion
when`ExtensionLoader` loads `CarMaker`'s  implementation `RaceCarMaker` ，the method `setWheelMaker` needs paramType `WheelMaker` which is also a SPI, It will be automatically loaded .
```

}
```

> for the method bind(), Adaptive will firstly search `server` key, if no Key were founded then will search `transport` key, to determine the implementation that the proxy represent for.

Reviewer: ```suggestion
For the method `bind()`, Adaptive will firstly search `server` key, if no key was found, it will search `transport` key, to determine the implementation which proxy represent for.
```

Author: The `Key` represent for `Key` in `Dubbo use URL (containing Key-Value) to pass the configuration.`, so it should be `Key`.

Reviewer: > The `Key` represent for `Key` in `Dubbo use URL (containing Key-Value) to pass the configuration.`, so it should be `Key`.

I am very sorry I was careless, thank you for correcting

}
```

Or:

```java
import org.apache.dubbo.common.extension.Activate;
import org.apache.dubbo.rpc.Filter;
 
@Activate("xxx") // when configed xxx parameter and the parameter has a valid value,the SPI is activated, for example configed cache="lru", auto acitivate CacheFilter.
public class XxxFilter implements Filter {
    // ...
}
```

Or:

```java
import org.apache.dubbo.common.extension.Activate;
import org.apache.dubbo.rpc.Filter;
 
@Activate(group = "provider", value = "xxx") // only activate for provider, group can be "provider" or "consumer"
public class XxxFilter implements Filter {
    // ...
}
```

> [^1]: Note: The config file here is in you own jar file, not in dubbo release jar file, Dubbo will scan all jar files for the same filename in ClassPath and merge together then

Reviewer: ```suggestion
[^1]: Note: The config file here is in you own jar file, not in dubbo release jar file, Dubbo will scan all jar files with the same filename in classpath and then merge them together
```
